File: JavaScript/Behavioral/Memento/Example1/Originator.js
// Originator.js

const Memento = require('./Memento');

// Originator creates and restores mementos
class Originator {
    constructor() {
        this.state = ""; // current state
    }

    setState(state) {
        this.state = state; // change internal state
        console.log(`Originator: setting state to "${state}"`);
    }

    getState() {
        return this.state; // read current state
    }

    save() {
        console.log(`Originator: saving to memento`);
        return new Memento(this.state); // create a memento
    }

    restore(memento) {
        this.state = memento.getState(); // restore from memento
        console.log(`Originator: state restored to "${this.state}"`);
    }
}

module.exports = Originator;

File: TheRayCode/CSharp/Behavioral/Memento/Example/Memento/Memento/Originator.cs
namespace Memento
{
    public class Originator
    {
        private string _state;

        // Set a new state
        public void SetState(string state)
        {
            _state = state;
        }

        // Get current state
        public string GetState()
        {
            return _state;
        }

        // Save state to memento
        public Memento SaveStateToMemento()
        {
            return new Memento(_state);
        }

        // Restore state from memento
        public void GetStateFromMemento(Memento memento)
        {
            _state = memento.GetState();
        }
    }
}